internal/testintegration/setup.go

```diff
@@ -110,85 +107,140 @@ func EnvTestSetup() (func(), error) {
-		return err
+	return th, nil
+}
+
+// EnvTestHarness enables integration tests to control the lifecycle of the
+// operator's controller-manager.
+type EnvTestHarness struct {
+
+	// Mgr is the manager
+	Mgr ctrl.Manager
+
+	// Client is the kubernetes client.
+	Client client.Client
+
+	// The actual EnvTest environment
+	testEnv *envtest.Environment
+
+	// testEnvCancel is the context cancel function for the testEnv
+	testEnvCancel context.CancelFunc
+
+	// cancel is the context cancel function for the manager
+	cancel context.CancelFunc
+
+	// stopped channel is closed when the manager actually stops
+	stopped chan int
+
+	// cfg is the client configuration from envtest
+	cfg *rest.Config
+
+	// s is the client scheme
+	s          *runtime.Scheme
+	testEnvCtx context.Context
+}
+
+// Teardown closes the TestEnv environment at the end of the testcase.
+func (h *EnvTestHarness) Teardown() {
+	h.testEnvCancel()
+	err := h.testEnv.Stop()
+	if err != nil {
+		Log.Error(err, "unable to stop envtest environment %v")
 	}
+}
+
+// StopManager stops the controller manager and waits for it to exit, returning an
+// error if the controller manager does not stop within 1 minute.
+func (h *EnvTestHarness) StopManager() error {
+	h.cancel()
+
+	// reassign cancel to do nothing, now that we've already
+	// called it.
+	h.cancel = func() {}
 
-	return teardownFunc, nil
+	select {
+	case <-h.stopped:
+		return nil
+	case <-time.After(1 * time.Minute):
+		return errors.New("manager did not stop after 1 minute")
+	}
 }
 
-func startManager(ctx context.Context, cfg *rest.Config, s *runtime.Scheme, proxyImage string) (context.CancelFunc, error) {
-	ctx, managerCancelFunc := context.WithCancel(ctx)
+// StartManager starts up the manager, configuring it with the proxyImage.
+func (h *EnvTestHarness) StartManager(proxyImage string) error {
+	var ctx context.Context
+	ctx, h.cancel = context.WithCancel(h.testEnvCtx)
 
 	// start webhook server using Manager
-	o := &testEnv.WebhookInstallOptions
-	mgr, err := ctrl.NewManager(cfg, ctrl.Options{
-		Scheme:             s,
+	o := &h.testEnv.WebhookInstallOptions
+	mgr, err := ctrl.NewManager(h.cfg, ctrl.Options{
+		Scheme:             h.s,
 		Host:               o.LocalServingHost,
 		Port:               o.LocalServingPort,
 		CertDir:            o.LocalServingCertDir,
 		LeaderElection:     false,
 		MetricsBindAddress: "0",
 	})
 	if err != nil {
-		return managerCancelFunc, fmt.Errorf("unable to start kuberenetes envtest %v", err)
+		return fmt.Errorf("unable to start manager %v", err)
 	}
+	h.Mgr = mgr
 
+	// Initialize the controller-runtime manager.
 	err = controller.SetupManagers(mgr, "cloud-sql-proxy-operator/dev", proxyImage)
-
 	if err != nil {
-		return managerCancelFunc, fmt.Errorf("unable to start kuberenetes envtest %v", err)
+		return fmt.Errorf("unable to start kuberenetes envtest %v", err)
 	}
 
+	// Run the manager in a goroutine, close the channel when the manager exits.
+	h.stopped = make(chan int)
```

Comment:
Can this be moved into the initializer?

Reply:
No. Tests can start and stop the manager. Whenever a test starts the manager, EnvTestHarness needs to make a new channel. When the manager is stopped, the channel gets closed and can't be reused.

Reply:
Got it. In that case, should these tests be threadsafe? If not, we should add a note in the initializer that this test harness is not threadsafe.

Reply:
I think I'll add a mutex so that StartManager() and StopManager() are thread safe. That seems easier than documenting that these need to be called from the same thread that instantiated the EnvTestHarness